cmd/fs_handlers.go

```diff
@@ -47,39 +46,43 @@ func handleStage(ctx *cli.Context, ctl *client.Client) error {
 		return fmt.Errorf("Failed to retrieve absolute path: %v", err)
 	}
 
-	info, err := os.Stat(absLocalPath)
+	_, err = os.Stat(absLocalPath)
 	if err != nil {
 		return err
 	}
 
-	if info.IsDir() {
-		return handleStageDirectory(ctx, ctl, absLocalPath, repoPath)
-	}
-
-	if !info.Mode().IsRegular() {
-		return fmt.Errorf("not adding non-regular file: %s", absLocalPath)
-	}
+	return handleStageDirectory(ctx, ctl, absLocalPath, repoPath)
+}
 
-	return ctl.Stage(absLocalPath, repoPath)
+type stagePair struct {
+	local, repo string
 }
 
-func handleStageDirectory(ctx *cli.Context, ctl *client.Client, root, repoRoot string) error {
-	// First create all directories:
-	type stagePair struct {
-		local, repo string
+func walk(root, repoRoot string, depth int) ([]stagePair, error) {
+	toBeStaged := make([]stagePair, 0)
+	toBeDereferenced := make([]stagePair, 0)
+	depth++
+	if depth > 255 {
+		return toBeStaged, fmt.Errorf("Exceeded allowed dereferencing depth")
 	}
-
-	toBeStaged := []stagePair{}
-
 	root = filepath.Clean(root)
 	repoRoot = filepath.Clean(repoRoot)
-
 	err := filepath.Walk(root, func(childPath string, info os.FileInfo, err error) error {
 		repoPath := filepath.Join("/", repoRoot, childPath[len(root):])
 
-		if info.IsDir() {
-			if err := ctl.Mkdir(repoPath, true); err != nil {
-				return e.Wrapf(err, "mkdir: %s", repoPath)
+		if info.Mode()&os.ModeSymlink != 0 {
+			resolvedPath, err := filepath.EvalSymlinks(childPath)
+			if err != nil {
+				return fmt.Errorf("Failed to resolve: %v: %v", childPath, err)
+			}
+			info, err = os.Stat(resolvedPath)
```

Comment:
Another evil thing: Symbolic links can resolve to more symbolic links. They're really a hostile concept to implement.